packages/chain-bridge/src/adapters/LayerZeroAdapter.sol

```diff
@@ -391,4 +702,96 @@ contract LayerZeroAdapter is
     }
+
+    /**
+     * @notice Handles protocol token fee payment flow for OFT transfers
+     * @dev This function manages the complete protocol token payment process for OFT transfers including:
+     *      - Fee quotation from OFT with token payment
+     *      - Token collection from the fee payer
+     *      - Allowance management for the OFT contract
+     *      - OFT sending with token fees
+     * @param operationId The operation ID for this transfer
+     * @param params ExecuteTransferParams from the bridge operation
+     * @param oft The OFT contract address
+     * @param sendParam Prepared send parameters for OFT
+     */
+    function _handleOFTProtocolTokenPayment(
+        bytes32 operationId,
+        BridgeTypes.ExecuteTransferParams calldata params,
+        address oft,
+        SendParam memory sendParam,
+        BridgeTypes.BridgeOptions calldata options
+    ) internal {
+        // Quote with token payment enabled
+        MessagingFee memory fee = IOFT(oft).quoteSend(sendParam, true);
+
+        // Validate that native fee is zero when paying in token
+        if (fee.nativeFee != 0) {
+            revert InsufficientFee(0, fee.nativeFee);
+        }
+
+        uint256 tokenFeeRequired = fee.lzTokenFee;
+
+        // Validate that user-provided fee amount matches quoted fee
+        if (options.feeTokenAmount != tokenFeeRequired) {
+            revert InsufficientFee(tokenFeeRequired, options.feeTokenAmount);
+        }
+
+        // Collect protocol token fee from the caller
+        _collectProtocolTokenFee(
+            operationId,
+            params.refundAddress,
+            tokenFeeRequired
+        );
+
+        // Ensure sufficient allowance for the OFT contract
+        _ensureSufficientAllowance(tokenFeeRequired, oft);
+
+        // Send with token payment (no native value required)
+        IOFT(oft).send{value: 0}(
+            sendParam,
+            MessagingFee(0, tokenFeeRequired),
+            params.refundAddress
+        );
+
+        // Refund any provided native value fully (since nativeFee == 0)
+        _refundNative(params.refundAddress, msg.value);
+
+        // Emit protocol fee spent event
+        emit ProtocolFeeSpent(operationId, protocolFeeToken, tokenFeeRequired);
+    }
+
+    /**
+     * @notice Handles native fee payment flow for OFT transfers
+     * @dev This function manages the native token payment process for OFT transfers
+     * @param params ExecuteTransferParams from the bridge operation
+     * @param oft The OFT contract address
+     * @param sendParam Prepared send parameters for OFT
+     */
+    function _handleOFTNativePayment(
+        BridgeTypes.ExecuteTransferParams calldata params,
+        address oft,
+        SendParam memory sendParam
+    ) internal {
+        // Quote with native payment
+        MessagingFee memory fee = IOFT(oft).quoteSend(sendParam, false);
+
+        // Validate sufficient native fee
+        if (msg.value < fee.nativeFee) {
```

Comment:
it should take into account `options.msgValue`
this should hold
`msg.value > fee.nativeFee + options.msgValue`

and its probably missing in another places